src/chain/chain_accumulator.rs

```diff
@@ -9,8 +9,66 @@
 use super::{AccumulatingEvent, NetworkEvent, Proof, ProofSet, SectionInfoSigPayload};
 use crate::id::PublicId;
 use log::LogLevel;
-use std::collections::{BTreeMap, BTreeSet};
-use std::mem;
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::{mem, rc::Rc};
+
+/// The threshold (number of unvoted votes) a node to be considered as unresponsive.
+pub const UNRESPONSIVE_THRESHOLD: usize = 48;
+/// The period (X consensued observations) after which node be considered as unresponsive.
+pub const UNRESPONSIVE_WINDOW: usize = 64;
+
+#[derive(Default)]
+struct VoteStatuses {
+    tracked_events: VecDeque<Rc<AccumulatingEvent>>,
+    unvoted: BTreeMap<PublicId, BTreeSet<Rc<AccumulatingEvent>>>,
+}
+
+impl VoteStatuses {
+    fn add_expectation(
+        &mut self,
+        event: AccumulatingEvent,
+        non_voters: BTreeSet<PublicId>,
+        all_members: &BTreeSet<PublicId>,
+    ) {
+        let event_rc = Rc::new(event);
+        for id in non_voters {
+            let events = self.unvoted.entry(id).or_insert_with(BTreeSet::new);
+            let _ = events.insert(event_rc.clone());
+        }
+        self.tracked_events.push_back(event_rc);
+
+        // Pruning old events
+        if self.tracked_events.len() > UNRESPONSIVE_WINDOW {
+            if let Some(removed_event) = self.tracked_events.pop_front() {
+                for events in self.unvoted.values_mut() {
+                    let _ = events.remove(&removed_event);
```

Comment:
Might be a good idea to clear the entry from `unvoted` if the set of unvoted events becomes empty.

Reply:
Looking at the code it does not seem clearer now that we do this extra work to remove the entry.
Why are we doing it?

Reply:
The idea was not to keep empty entries unnecessarily - they probably don't use much memory, but they use some and would do so without purpose. So I think we should clean this up, but another option is to leave this in place and let the mechanism for cleaning old peers handle it when it kicks in.

Reply:
first need to remove the `old` unvoted events, then as suggested to remove peer that voted all.
I could do a 
```
self.unvoted = mem::replace(&mut self.unvoted, BTreeMap::new())
                .into_iter()
                .filter(|(_, events)| !events.is_empty())
                .collect();
```
outside, which is not super efficient but clearer to read?

Reply:
> they probably don't use much memory, but they use some and would do so without purpose.

The problem is we now have more complicated code, but also many more addition removal to that map, which is more performance consuming. Additionally, the previous mechanism ensured we only keep the needed members, so always consumed the elder size (i.e 7 entries).

It feels a lot more convoluted now, while the gain of removing the entry seem really unclear, and probably unneeded.

Reply:
>  which is not super efficient but clearer to read?
We need to keep the code efficient (as we seen, it is easy to get a big test performance drop).